src/fortrancode.l

```diff
@@ -938,25 +987,37 @@ LANGUAGE_BIND_SPEC BIND{BS}"("{BS}C{BS}(,{BS}NAME{BS}"="{BS}"\""(.*)"\""{BS})?")
                                           yy_pop_state();
                                           YY_FTN_RESET
                                         }
-<Start>"end"{BS}("block"{BS}"data"|{SUBPROG}|"module"|"program"|"enum"|"type"|"interface")?{BS}     {  // Fortran subroutine or function ends
-                                          //cout << "===> end function " << yytext << endl;
+<Start>^{BS}"end"{BS}/(\n|!|;)          { // bare end keyword.
+                                          //cout << "===> bare end keyword " << yytext << endl;
+                                          //printf("===> line number %d bare end keyword |%s|\n",g_yyLineNr,yytext);
                                           endScope();
    					  startFontClass("keyword");
   					  codifyLines(yytext);
 					  endFontClass();
-                                          yy_push_state(YY_START);
-                                          BEGIN(Subprogend);
                                         }
-<Subprogend>{ID}/{BS}(\n|!|;)             {
-					  generateLink(*g_code,yytext);
-                                          yy_pop_state();
+<Start>"end"{BS}("block"{BS}"data"|{SUBPROG}|"module"|"program"|"enum"|"type"|"interface"){BS}/(\n|!|;) {
+                                          // Fortran subroutine or function ends, without name.
+                                          //cout << "===> end function w/o name " << yytext << endl;
+                                          //printf("===> line number %d end function w/o name |%s|\n",g_yyLineNr,yytext);
+                                          endScope();
+   					  startFontClass("keyword");
+  					  codifyLines(yytext);
+					  endFontClass();
                                         }
-<Start>"end"{BS}("block"{BS}"data"|{SUBPROG}|"module"|"program"|"enum"|"type"|"interface"){BS}/(\n|!|;) {  // Fortran subroutine or function ends
-                                          //cout << "===> end function " << yytext << endl;
+<Start>"end"{BS}("block"{BS}"data"|{SUBPROG}|"module"|"program"|"enum"|"type"|"interface"){BS}/{ID}{BS}(\n|!|;) {
+                                          // Fortran subroutine or function ends, with name.
+                                          //cout << "===> end function w/ name " << yytext << endl;
+                                          //printf("===> line number %d end function w/ name |%s|\n",g_yyLineNr,yytext);
                                           endScope();
    					  startFontClass("keyword");
   					  codifyLines(yytext);
 					  endFontClass();
+                                          yy_push_state(YY_START);
+                                          BEGIN(Subprogend);
+                                        }
+<Subprogend>{ID}/{BS}(\n|!|;)             {
+					  generateLink(*g_code,yytext);
+                                          yy_pop_state();
```

Comment:
I don't see the advantage of the new construct in splitting everything in different parts compared to the old construct.

Reply:
{ID} should not be allowed to follow a bare "end" keyword.  In hindsight, I have updated the PR to fix that with minimal changes to the original construct and without additional splitting.